modules/vebal/root-gauges.onchain.ts

```diff
@@ -0,0 +1,221 @@
+} as const;
+
+export type RootGauge = {
+    gaugeAddress: Address;
+    network: Chain;
+    isKilled: boolean;
+    relativeWeight: number;
+    relativeWeightCap?: string;
+    recipient?: string;
+    stakingId?: Address;
+    isInSubgraph: boolean;
+};
+
+/**
+ *
+ * This is the first proof of concept using viem's multicall
+ * It contains ad-hoc helpers to make the code easier in this concrete scenario but in the future we will try to create a more generic
+ * multicaller implementation to generalize any multicall flow
+ *
+ */
+export class OnChainRootGauges {
+    constructor(
+        private publicClient: PublicClient = mainnetNetworkConfig.publicClient!,
+        private readContract = publicClient.readContract,
+    ) {}
+
+    async getRootGaugeAddresses(): Promise<string[]> {
+        const totalGauges = Number(
+            await this.readContract({
+                ...gaugeControllerContract,
+                functionName: 'n_gauges',
+            }),
+        );
+        const addresses = await this.publicClient.multicall({
+            allowFailure: false,
+            contracts: this.gaugeControllerCallsByIndex(totalGauges, 'gauges'),
+        });
+
+        return addresses.map((address) => address.toLowerCase());
+    }
+
+    async fetchOnchainRootGauges(gaugeAddresses: Address[]): Promise<RootGauge[]> {
+        const totalGaugesTypes = Number(
+            await this.readContract({
+                ...gaugeControllerContract,
+                functionName: 'n_gauge_types',
+            }),
+        );
+
+        const typeNames = await this.publicClient.multicall({
+            allowFailure: false,
+            contracts: this.gaugeControllerCallsByIndex(totalGaugesTypes, 'gauge_type_names'),
+        });
+
+        const relativeWeights = await this.multicallGaugeController(gaugeAddresses, 'gauge_relative_weight');
+
+        /*
+            gauge_types are not reliable because they are manually input by Maxis
+            We will use subgraph chain field instead
+            However, we keep pulling this gauge_types cause they can be useful for debugging (when a root gauge is not found in the subgraph)
+        */
+        const gaugeTypeIndexes = await this.multicallGaugeController(gaugeAddresses, 'gauge_types');
+        const gaugeTypes = mapValues(gaugeTypeIndexes, (type) => typeNames[Number(type)]);
+
+        const isKilled = await this.multicallRootGauges(gaugeAddresses, 'is_killed');
+
+        const relativeWeightCapsWithFailures = await this.multicallRootGaugesAllowingFailures(
+            gaugeAddresses,
+            'getRelativeWeightCap',
+        );
+        const relativeWeightCaps = mapValues(relativeWeightCapsWithFailures, (r) => r.result);
+
+        let rootGauges: RootGauge[] = [];
+        gaugeAddresses.forEach((gaugeAddress) => {
+            const relativeWeight = relativeWeightCaps[gaugeAddress];
+            if (gaugeTypes[gaugeAddress] === 'Liquidity Mining Committee') return;
```

Comment:
What is this?

Reply:
It is an special root gauge: https://etherscan.io/address/0x7aa5475b2ea29a9f4a1b9cf1cb72512d1b4ab75e#readContract 

with type 0 (Liquidity Mining Committee):

<img width="1210" alt="LiquidityMinningComitee" src="https://github.com/beethovenxfi/beethovenx-backend/assets/1316240/ee45d962-3f05-4b27-a093-721bdbe25ae2">

We just exclude it.